platform/pulp/platform/models/progress.py

```diff
@@ -0,0 +1,195 @@
+
+    All progress reports have a message, state, and are related to a Task.
+
+    Fields:
+
+    :cvar message: A short message for the progress update, typically shown to the user. (required)
+    :type message: models.TextField
+
+    :cvar state: The state of the progress update. Defaults to `WAITING`. This field uses a limited
+                 set of choices of field states. See `STATES` for possible states.
+    :type state: models.TextField
+
+    Relations:
+
+    :cvar task: The task associated with this progress report. If left unset when save() is called
+                it will be set to the current task_id.
+    :type task: models.ForeignKey
+    """
+    WAITING = 'waiting'
+    SKIPPED = 'skipped'
+    RUNNING = 'running'
+    FINISHED = 'finished'
+    ERROR = 'error'
+    CANCELED = 'canceled'
+    STATES = (
+        (WAITING, 'Waiting'),
+        (SKIPPED, 'Skipped'),
+        (RUNNING, 'Running'),
+        (FINISHED, 'Finished'),
+        (ERROR, 'Error'),
+        (CANCELED, 'Canceled')
+    )
+    message = models.TextField()
+    state = models.TextField(choices=STATES, default=WAITING)
+    task = fields.GenericRelation(Task)
+
+    class Meta:
+        abstract = True
+
+    def call_and_observe(self, func, args, kwargs):
+        """
+        Call func() and handle progress reporting changes for the RUNNING FINISHED and ERROR states.
+
+        This method effectively calls:
+
+            return func(*args, **kwargs)
+
+        The progress report state is saved as RUNNING just before calling func(). If an exception
+        occurs, the progress report state is saved as ERROR and the exception is raised. If no
+        exception occurs, the progress report state is saved as FINISHED.
+
+        :param f: The function to be called.
+        :type f:  function
+
+        :param args: The positional arguments to be passed to func
+        :type args:  list
+
+        :param kwargs: The keyword arguments to be passed to func
+        :type kwargs:  dict
+
+        :return: The return value of func(*args, **kwargs)
+        """
+        try:
+            if self.state != self.RUNNING:
+                self.state = self.RUNNING
+                self.state.save()
+            return func(*args, **kwargs)
+        except Exception:
+            self.state = self.ERROR
+            self.save()
+            raise
+        finally:
+            self.state = self.FINISHED
+            self.save()
+
+
+class ProgressSpinner(ProgressReportingBase):
+    """
+    Shows progress reporting when the count of items is not known.
+
+    Plugin writers should create these objects to show progress reporting of a single step or
+    aspect of work which has a name and a state. For example:
+
+    ProgressSpinner(message='Publishing Metadata') # default state is 'waiting'
+
+    ProgressSpinner(message='Publishing Metadata', state='running') # specify 'running' instead
+
+    # Update the state to FINISHED and save it
```

Comment:
This is a great idea. I've adjusted all of my docstring examples to this. Thanks @jeremycline